File: training-data/scala/Trie.scala
class TrieNode {
  val children = scala.collection.mutable.Map[Char, TrieNode]()
  var isEndOfWord = false
}

class Trie {
  private val root = new TrieNode

  def insert(word: String): Unit = {
    var node = root
    for (ch <- word) {
      node = node.children.getOrElseUpdate(ch, new TrieNode)
    }
    node.isEndOfWord = true
  }

  def search(word: String): Boolean = {
    var node = root
    for (ch <- word) {
      node.children.get(ch) match {
        case Some(next) => node = next
        case None => return false
      }
    }
    node.isEndOfWord
  }
}

object TrieDemo {
  def main(args: Array[String]): Unit = {
    val trie = new Trie
    List("cat", "car", "card").foreach(trie.insert)
    println(trie.search("car"))
    println(trie.search("ca"))
  }
}

File: training-data/swift/trie.swift
final class TrieNode {
    var children: [Character: TrieNode] = [:]
    var isEndOfWord = false
}

final class Trie {
    private let root = TrieNode()

    func insert(_ word: String) {
        var node = root
        for ch in word {
            if node.children[ch] == nil {
                node.children[ch] = TrieNode()
            }
            node = node.children[ch]!
        }
        node.isEndOfWord = true
    }

    func search(_ word: String) -> Bool {
        var node = root
        for ch in word {
            guard let next = node.children[ch] else { return false }
            node = next
        }
        return node.isEndOfWord
    }
}

let trie = Trie()
["cat", "car", "card"].forEach { trie.insert($0) }
print(trie.search("car"))
print(trie.search("ca"))
